Back up and restore SecurityDescriptors on Windows

aneesh-n opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
Restic did not back up SecurityDescriptors of files on Windows.
Restic now backs up and restores SecurityDescriptors (which includes owner, group, discretionary access control list (DACL), system access control list (SACL)) when backing up files and folders on Windows. This requires the user to be a member of backup operators or the application must be run as admin.
If that is not the case, only the current user's owner, group and DACL will be backed up and during restore only the DACL of the backed file will be restored while the current user's owner and group will be set during the restore.

We may want to add a backup/restore flag for disabling Security descriptor for Windows as there may be valid cases where users do not want to backup/restore security descriptors of files in Windows. 
This can be done as a separate PR.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Follow-up to https://github.com/restic/restic/pull/4611


Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].
-->

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

aneesh-n committed:
Create helper for SecurityDescriptor related functions

aneesh-n committed:
Fix CombineErrors and fillExtendedAttr error handling

aneesh-n committed:
Add support for backup/restore of security descriptors

aneesh-n committed:
Add test cases for security descriptors

aneesh-n committed:
Update docs for security descriptors

aneesh-n committed:
Add changelog and fix lint error

aneesh-n commented on internal/fs/sd_windows.go:
Would have to move this helper into another PR to reduce the total changed lines for this PR.

aneesh-n committed:
Clean up SecurityDescriptor helper

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -114,7 +117,7 @@ func (s statT) mtim() syscall.Timespec {
 func (s statT) ctim() syscall.Timespec {
 	// Windows does not have the concept of a "change time" in the sense Unix uses it, so we're using the LastWriteTime here.
-	return syscall.NsecToTimespec(s.LastWriteTime.Nanoseconds())
+	return s.mtim()
```
This fixes func statT.mtim is unused (U1000)go-staticcheck and removes code duplication

aneesh-n commented on internal/restic/node.go:
```diff
@@ -719,12 +721,7 @@ func (node *Node) fillExtra(path string, fi os.FileInfo) error {
-			err = errEx
-		} else {
-			debug.Log("Error filling extended attributes for %v at %v : %v", node.Name, path, errEx)
-		}
```
Error from fillExtendedAttributes was getting eaten up after this code was updated in https://github.com/restic/restic/pull/4611/files#diff-ed902b2edb54dfe5e342c38a7e7d122cacc6b1e84e956dfc8689b5efd7922fd4

aneesh-n replied:
I can move this to the next PR for extended attrs if needed.

MichaelEischer reviewed (changes requested):
Sorry for taking so long to review this PR, somehow it always ended up in my "it would take too long to review the PR today" category :-( . I've added quite a few comments below, although most of them are just nitpicks.

The only larger issue is the encoding of the SECURITY_DESCRIPTOR struct. I don't think it's a good idea to store the raw data structure. Instead, I think the PR should use the SDDL encoding, see the comment for more details.

aneesh-n committed:
Merge branch 'master' into windows-securitydesc

aneesh-n commented on internal/fs/sd_windows_test.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/restic/restic/internal/test"
+)
+
+func Test_SetGetFileSecurityDescriptors(t *testing.T) {
```
Done

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+	return string(utf16.Decode(displayNameBuffer[:displayBufSize]))
+}
+
+func adjustTokenPrivileges(token windows.Token, releaseAll bool, input *byte, outputSize uint32, output *byte, requiredSize *uint32) (success bool, err error) {
```
Probably, these were copied from winio.

MichaelEischer replied:
I've found the source file: https://github.com/microsoft/go-winio/blob/main/zsyscall_windows.go . _I don't want to see copied code ever again that does not clearly state where it originates from._ It's just ridiculous that I have to search for the original source of these functions.

Omitting that information also forces me to waste time on reviewing known good code.

To make matter worse, there have been a few small changes to those functions (apparently to fix linter warnings), which when compared to the latest upstream version of the autogenerated code seem to be buggy. See my other comments for details.

aneesh-n replied:
That's a little harsh. I wrote this code 2 years ago before I decided to submit it back to restic.

The upstream code was changed only last month - https://github.com/microsoft/go-winio/commit/3c9576c9346a1892dee136329e7e15309e82fb4f
Eg.
	- r0, _, _ := syscall.Syscall(procHcsFormatWritableLayerVhd.Addr(), 1, uintptr(handle), 0, 0)
	+ r0, _, _ := syscall.SyscallN(procHcsFormatWritableLayerVhd.Addr(), uintptr(handle))
	
The changes I made 2 years ago were to fix the deprecations - 
Deprecated: Use SyscallN instead.	

Extra 0s at the end of SyscallN are ignored and hence this slip wasn't caught, and it also doesn't cause any bugs, though I agree it should be cleaned.

Also, I get your point on including links to the source and will try to do that, this was missed because the code was created from upstream a long time ago.

MichaelEischer replied:
> That's a little harsh.

Let's just say that I was rather annoyed after noticing that I have to diff the syscall wrappers and search for the corresponding source code.

> I wrote this code 2 years ago before I decided to submit it back to restic.

That perfectly explains why the code looks the way it does. Thanks for contributing those changes!

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+}
+
+// DisableProcessPrivileges disables privileges globally for the process.
+func DisableProcessPrivileges(names []string) error {
```
Had kept these incase we need to disable externally based on flags or something. It was unused, I've removed them now.

MichaelEischer replied:
The function still seems to exist.

aneesh-n replied:
DisableBackupPrivileges and DisableRestorePrivileges were removed. This one got left out.
Removed now.

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+		s += getPrivilegeName(p)
+		s += `"`
+	}
+	if backupPrivilegeError != nil {
```
Fixed

MichaelEischer replied:
using `os.OpenFile` should be fine. By now, Go should also handle most cases on its own that once required `fixpath(...)`. And I guess it's unlikely that users input some crazy path when trying to get debug logs from restic.

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+	return member, nil
+}
+
+// The code below was adapted from github.com/Microsoft/go-winio under MIT license.
```
Done

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+
+// SecurityDescriptorBytesToStruct converts the security descriptor bytes representation
+// into a pointer to windows SECURITY_DESCRIPTOR.
+func SecurityDescriptorBytesToStruct(sd []byte) (*windows.SECURITY_DESCRIPTOR, error) {
```
Moved them up

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+// Do the interface allocations only once for common
+// Errno values.
+const (
+	errnoErrorIOPending = 997
```
This is from windows - https://github.com/microsoft/go-winio/blob/3c9576c9346a1892dee136329e7e15309e82fb4f/tools/mkwinsyscall/mkwinsyscall.go#L948

aneesh-n commented on changelog/unreleased/pull-4708:
```diff
@@ -0,0 +1,12 @@
+Enhancement: Back up and restore SecurityDescriptors on Windows
+
+Restic did not back up SecurityDescriptors of files on Windows.
```
Done

aneesh-n commented on doc/040_backup.rst:
```diff
@@ -481,12 +481,17 @@ written, and the next backup needs to write new metadata again. If you really
 
+Backing up full security descriptors on windows is only possible when the user 
+has ``SeBackupPrivilege``privilege or is running as admin. This is a restriction 
+of windows not restic.
```
Done

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -23,6 +23,9 @@ type WindowsAttributes struct {
 	// FileAttributes is used for storing file attributes for windows files.
 	FileAttributes *uint32 `generic:"file_attributes"`
+	// SecurityDescriptor is used for storing security descriptors which includes
+	// owner, group, discretionary access control list (DACL), system access control list (SACL))
```
Fixed

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+
+// DisableBackupPrivileges disables privileges that are needed for backup operations.
+// They may be reenabled if GetSecurityDescriptor is called again.
+func DisableBackupPrivileges() error {
```
I was thinking these would be useful helpers in future, but I've removed them now.

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+	if err != nil {
+		return false, errors.Errorf("sid error: %s", err)
+	}
+	token := windows.Token(0)
```
windows.Token(0) directly creates a token object representing the access token of the current process with a specific access level. It's a direct way to obtain a token without explicitly opening a handle to the process token.
https://github.com/golang/go/issues/28804

Also, I have tested this.

MichaelEischer replied:
Thanks for the link.

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+}
+
+// IsAdmin checks if current user is an administrator.
+func IsAdmin() (isAdmin bool, err error) {
```
Done

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+
+// SecurityDescriptorBytesToStruct converts the security descriptor bytes representation
+// into a pointer to windows SECURITY_DESCRIPTOR.
+func SecurityDescriptorBytesToStruct(sd []byte) (*windows.SECURITY_DESCRIPTOR, error) {
```
In Microsoft's go-winio package (check https://github.com/microsoft/go-winio/blob/bc421d9108ade8ae2f76c8ef3e90cd4ee690e2c0/backuptar/tar.go#L130), they have deprecated sddl-based usage of SecurityDescriptors and instead use base64 encoded raw descriptors. 

I had updated my implementation from sddl to raw accordingly.
"Maintaining old SDDL-based behavior for backward compatibility. All new tar headers written by this library will have raw binary for the security."

Refer - https://github.com/microsoft/go-winio/commit/0ab2de7e65cca04fcdd6bc8846b0d6e6f2732bd7
"Remove use of SDDL in tar headers
Switching to using straight binary instead of SDDL in tar headers to avoid failures in converting domain SIDs into/out of SDDL."

While testing SDs, they were not getting restored for domain accounts when using SSDL and it worked properly only after updating it to use raw binary as specified in go-winio.

MichaelEischer replied:
> I guess SDs won't work when moving between 32 and 64 bit systems.

According to https://nostarch.com/download/WindowsSecurityInternals_Chapter5.pdf the self-relative SDs, which are used here, are encoded the same on 32/64 bit systems. That also matches that https://learn.microsoft.com/en-us/windows/win32/secauthz/absolute-and-self-relative-security-descriptors says that the format can be stored/transmitted.

So I'm not entirely happy with how we have to serialize the SDs. But as the code is based on https://github.com/microsoft/go-winio/blob/3c9576c9346a1892dee136329e7e15309e82fb4f/sd.go#L118 , the approach is probably fine.

aneesh-n committed:
Fix review comments

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+	}
+	if err != nil {
+		if isHandlePrivilegeNotHeldError(err) {
+			lowerPrivileges = true
```
If there are concurrent calls, they will all try with the higher privilege and fallback to testing with getNamedSecurityInfoLow and eventually all calls will use either lower or higher privilege.
Is sync really needed for every call for this?

MichaelEischer replied:
I agree that it probably won't break anything. However, the golang race detector will probably complain about some of that code. The simplest solution would be to use an atomic boolean instead: https://pkg.go.dev/sync/atomic#Bool .

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+		sd, err = getNamedSecurityInfoHigh(sd, err, filePath)
+	}
+	if err != nil {
+		if isHandlePrivilegeNotHeldError(err) {
```
Even if it can, I guess there's no point in the fallback for this case. 
I will add - !lowerPrivileges && isHandlePrivilegeNotHeldError(err)

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -12,10 +13,66 @@ import (
+	expectedNode := getNode(fileName, fileType, genericAttributes)
+
+	// Restore the file/dir and restore the meta data including the security descriptors.
+	testPath, node := restoreAndGetNode(t, tempDir, expectedNode, false)
```
Yes, I considered this. I did test that it does not cause conflicts.
This case should be covered under node_windows_test - TestRestoreSecurityDescriptors right?

MichaelEischer replied:
Yeah, that test should notice if the chmod would interfere with the security descriptors.

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,469 @@
+)
+
+// Flags for backup and restore with admin permissions
+var highSecurityFlags windows.SECURITY_INFORMATION = windows.OWNER_SECURITY_INFORMATION | windows.GROUP_SECURITY_INFORMATION | windows.DACL_SECURITY_INFORMATION | windows.SACL_SECURITY_INFORMATION | windows.LABEL_SECURITY_INFORMATION | windows.ATTRIBUTE_SECURITY_INFORMATION | windows.SCOPE_SECURITY_INFORMATION | windows.BACKUP_SECURITY_INFORMATION | windows.PROTECTED_DACL_SECURITY_INFORMATION | w ...
```
According to docs it does seem like BACKUP_SECURITY_INFORMATION should cover all, but on removing those other flags like PROTECTED_DACL_SECURITY_INFORMATION, TestSetGetFileSecurityDescriptors fails saying the input and output DACLs do not match. If the flags are added back, the DACLs are restored properly.

The unprotected flags are for acls not protected from being modified by inheritance, which are for auditing purposes.
I will add those flags as well.

aneesh-n replied:
Do you think we should have a flag to be able to disable backup of security attributes? 
I think there may be some valid cases when users want to backup files and just have default security attributes used when restored.

MichaelEischer replied:
I don't see much reason to not include the security attributes in the backup. But we could discuss about a `restore` option.

aneesh-n replied:
Agree, it makes sense to add and option to include/exclude SDs during restore. This can be added in this PR or added as a new PR.

MichaelEischer replied:
let's keep that option out of this PR. It's already large enough as it is ;-)

aneesh-n committed:
Fix review comments for privileges and security flags

MichaelEischer commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,431 @@
+	SE_PRIVILEGE_ENABLED = windows.SE_PRIVILEGE_ENABLED
+
+	//revive:disable-next-line:var-naming ALL_CAPS
+	ERROR_NOT_ALL_ASSIGNED syscall.Errno = windows.ERROR_NOT_ALL_ASSIGNED
```
The upstream version uses `windows.Errno`.

aneesh-n replied:
This deprecation clean-up was done upstream 9 months ago after this feature code was written. 
Have updated the code now

MichaelEischer commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,431 @@
+	}
+
+	if err != nil {
+		if isHandlePrivilegeNotHeldError(err) {
```
Should also check for `!lowerPrivileges`

aneesh-n replied:
Fixed

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,431 @@
+}
+
+func _lookupPrivilegeDisplayName(systemName *uint16, name *uint16, buffer *uint16, size *uint32, languageID *uint32) (err error) {
+	r1, _, e1 := syscall.SyscallN(procLookupPrivilegeDisplayNameW.Addr(), uintptr(unsafe.Pointer(systemName)), uintptr(unsafe.Pointer(name)), uintptr(unsafe.Pointer(buffer)), uintptr(unsafe.Pointer(size)), uintptr(unsafe.Pointer(languageID)), 0)
```
Fixed

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,431 @@
+}
+
+func _lookupPrivilegeName(systemName *uint16, luid *uint64, buffer *uint16, size *uint32) (err error) {
+	r1, _, e1 := syscall.SyscallN(procLookupPrivilegeNameW.Addr(), uintptr(unsafe.Pointer(systemName)), uintptr(unsafe.Pointer(luid)), uintptr(unsafe.Pointer(buffer)), uintptr(unsafe.Pointer(size)), 0, 0)
```
Fixed

aneesh-n commented on internal/fs/sd_windows.go:
```diff
@@ -0,0 +1,431 @@
+	}
+	r0, _, e1 := syscall.SyscallN(procAdjustTokenPrivileges.Addr(), uintptr(token), uintptr(_p0), uintptr(unsafe.Pointer(input)), uintptr(outputSize), uintptr(unsafe.Pointer(output)), uintptr(unsafe.Pointer(requiredSize)))
+	success = r0 != 0
+	if !success {
```
This windows api always returns an error even in case of success, warnings (partial success) and error cases.

Full success - When we call this with admin permissions, it returns DNS_ERROR_RCODE_NO_ERROR (0). This gets translated to errErrorEinval and ultimately in adjustTokenPrivileges, it gets ignored.

Partial success - If we call this api without admin privileges, privileges related to SACLs do not get set and though the api returns success, it returns an error - golang.org/x/sys/windows.ERROR_NOT_ALL_ASSIGNED (1300)
This gets propagated up the call hierarchy and then gets printed as a debug log in enableRestorePrivilege or enableBackupPrivilege. 
Now that we have changed the handling in those functions to print debug log, it will start printing debug logs even in case ERROR_NOT_ALL_ASSIGNED was thrown, which is a warning and in our case is expected in case we run restic without admin permissions. In that case adjustPrivileges will return a PrivilegeError even though other privileges other than SACL access were still set.

Instead of eating the warning on one extreme or throwing a Privilege error on the other extreme, I can handle this warning in adjustPrivileges and add a debug statement - 

debug.Log("Not all requested privileges were fully set: %v. AdjustTokenPrivileges returned warning: %v", privileges, err)
This debug statement will get logged when we call restic without admin permissions.

MichaelEischer replied:
The debug.Log is fine.

Judging from https://learn.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-adjusttokenprivileges `if true {` is in fact the correct variant (even though it looks weird. However, calling errnoErr doesn't really make sense in that case), as `GetLastError` can return either ERROR_SUCCESS or ERROR_NOT_ALL_ASSIGNED if adjustTokenPrivileges succeeded.

aneesh-n committed:
Fix review comments

Change lowerPrivileges from bool to atomic.Bool.
Add missing cleanup from upstream go-winio.
Add handling for ERROR_NOT_ALL_ASSIGNED warning.

MichaelEischer reviewed (approved):
LGTM. Thanks!

zumm commented:
Is there a way to disable this feature? I'm having tough times synchronizing files between machines.
